src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs

```diff
@@ -392,18 +396,34 @@ internal static void ConfigureAlpn(SslServerAuthenticationOptions serverOptions,
             return true;
         }
 
-        private bool RemoteCertificateValidationCallback(object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
-            RemoteCertificateValidationCallback(_options.ClientCertificateMode, _options.ClientCertificateValidation, certificate, chain, sslPolicyErrors);
+        private bool RemoteCertificateValidationCallback(object sender, X509Certificate? certificate, X509Chain? chain, SslPolicyErrors sslPolicyErrors)
+        {
+            Debug.Assert(_options != null, "Middleware must be created with options.");
+
+            return RemoteCertificateValidationCallback(_options.ClientCertificateMode, _options.ClientCertificateValidation, certificate, chain, sslPolicyErrors);
+        }
 
-        private SslDuplexPipe CreateSslDuplexPipe(IDuplexPipe transport, MemoryPool<byte> memoryPool)
+        private SslDuplexPipe CreateSslDuplexPipe(IDuplexPipe transport, MemoryPool<byte>? memoryPool)
         {
-            var inputPipeOptions = new StreamPipeReaderOptions
-            (
-                pool: memoryPool,
-                bufferSize: memoryPool.GetMinimumSegmentSize(),
-                minimumReadSize: memoryPool.GetMinimumAllocSize(),
-                leaveOpen: true
-            );
+            StreamPipeReaderOptions inputPipeOptions;
+
+            if (memoryPool != null)
+            {
+                inputPipeOptions = new StreamPipeReaderOptions
+                (
+                    pool: memoryPool,
+                    bufferSize: memoryPool.GetMinimumSegmentSize(),
+                    minimumReadSize: memoryPool.GetMinimumAllocSize(),
+                    leaveOpen: true
+                );
+            }
+            else
+            {
+                inputPipeOptions = new StreamPipeReaderOptions
+                (
+                    leaveOpen: true
+                );
```

Comment:
We should probably revert this. The StreamPipeReaderOptions ctor does allow a null memory pool to be specified and the GetMinimumSegmentSize() and GetMinimumAllocSize() are returning slightly non-default options when the memory pool is null. Maybe these shouldn't be extension methods so it's not so weird that they accept null.